### PR TITLE
Add types to dataset descriptions

### PR DESCRIPTION
**Pull request by jenshnielsen**
Makes dataset.description fully typed

### CODE REVIEW

File: qcodes/dataset/descriptions/dependencies.py [hunk 1]
```diff
@@ -488,7 +488,7 @@ def __repr__(self) -> str:
                f"standalones={self.standalones})")
         return rep
 
-    def __eq__(self, other):
+    def __eq__(self, other: Any) -> bool:
 
         def sorter(inp: Any) -> List[Any]:
```
Comment (astafan8) [inline qcodes/dataset/descriptions/dependencies.py]: shouldn't this one actually `ParamSpecBase`, not `Any`?

File: qcodes/dataset/descriptions/param_spec.py [hunk 2]
```diff
@@ -108,11 +108,11 @@ class ParamSpec(ParamSpecBase):
...
     def __init__(self, name: str,
                  paramtype: str,
-                 label: Optional[str]=None,
-                 unit: Optional[str]=None,
-                 inferred_from: Sequence[Union['ParamSpec', str]]=None,
-                 depends_on: Sequence[Union['ParamSpec', str]]=None,
-                 **metadata) -> None:
+                 label: Optional[str] = None,
+                 unit: Optional[str] = None,
+                 inferred_from: Sequence[Union['ParamSpec', str]] = None,
```
Comment (astafan8) [inline qcodes/dataset/descriptions/param_spec.py]: Optional[Sequence... ?

File: qcodes/dataset/descriptions/dependencies.py (hunk 1 above)
Comment (jenshnielsen) [reply qcodes/dataset/descriptions/dependencies.py]: Nope you are allowed to compare this with anything (but it returns False naturally)

File: qcodes/dataset/descriptions/param_spec.py (hunk 2 above)
Comment (jenshnielsen) [reply qcodes/dataset/descriptions/param_spec.py]: Yes both these types are apparently wrong as they are

File: qcodes/dataset/descriptions/dependencies.py (hunk 1 above)
Comment (astafan8) [reply qcodes/dataset/descriptions/dependencies.py]: i meant the `Any` in the `sorter` method

Comment (jenshnielsen) [reply qcodes/dataset/descriptions/dependencies.py]: OK :) I think this is correct

**Pull request merged**